In [2]:
import pandas as pd
import numpy as np

In [3]:
pd.set_option('display.max_columns', None)

## Reading the data files

In [4]:
df = pd.read_csv('./apps.csv')

## To get basic idea of the data set

In [5]:
# returns shape of the data file (rows,columns)
df.shape

(10841, 12)

In [6]:
# gives basic information like how many non-null values we have for each column(Feature) set and the type of the column.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  int64  
 4   Size_MBs        10841 non-null  float64
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content_Rating  10841 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last_Updated    10841 non-null  object 
 11  Android_Ver     10839 non-null  object 
dtypes: float64(2), int64(1), object(9)
memory usage: 1016.5+ KB


In [7]:
# returns statistical values for numerical Features
df.describe().transpose()

count           mean           std       min   25%     50%  \
Rating     9367.0       4.191513  5.157352e-01  1.000000   4.0     4.3   
Reviews   10841.0  444111.928051  2.927629e+06  0.000000  38.0  2094.0   
Size_MBs  10841.0      19.774147  2.140435e+01  0.008301   4.9    11.0   

              75%         max  
Rating        4.5         5.0  
Reviews   54768.0  78158306.0  
Size_MBs     27.0       100.0

In [8]:
# to get any 8 random rows
df.sample(8)

App             Category  Rating  Reviews  \
7675                 The translator                TOOLS     4.5    23805   
5804  KDRAMA Amino for K-Drama Fans               SOCIAL     4.6     4046   
6271      Golden Dictionary (FR-AR)  BOOKS_AND_REFERENCE     4.2     5775   
2987                  SMS Au revoir               FAMILY     NaN       17   
9318                Real Basketball               SPORTS     4.5  1605267   
9469   CM Security Lite - Antivirus                TOOLS     4.6   262076   
8505     GO Launcher EX UI5.0 theme      PERSONALIZATION     4.3   111634   
4281          EZ-GUI Ground Station                TOOLS     4.7     3696   

      Size_MBs    Installs  Type Price Content_Rating             Genres  \
7675       4.1   1,000,000  Free     0       Everyone              Tools   
5804      63.0     100,000  Free     0           Teen             Social   
6271       4.9     500,000  Free     0       Everyone  Books & Reference   
2987       1.7       5,000  Free     0       Everyone      Entertainment   
9318      14.0  10,000,000  Free     0       Everyone             Sports   
9469       1.2  10,000,000  Free     0       Everyone              Tools   
8505      13.0   5,000,000  Free     0       Everyone    Personalization   
4281      29.0      50,000  Free     0       Everyone              Tools   

          Last_Updated         Android_Ver  
7675   August 13, 2017        4.0.3 and up  
5804     July 13, 2018        4.0.3 and up  
6271     July 19, 2018          4.2 and up  
2987  January 15, 2018          4.0 and up  
9318  December 5, 2016  Varies with device  
9469   January 4, 2018          2.3 and up  
8505      May 26, 2014          2.0 and up  
4281   August 10, 2017          4.3 and up

In [9]:
# To get top five rows
df.head()

App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  \
0        0       8.7        0  Paid  $13.99           Teen           Social   
1        0      33.0        0  Paid   $2.99       Everyone        Education   
2        0       5.5        0  Paid   $1.49       Everyone  Personalization   
3        0      19.0        0   NaN       0   Everyone 10+         Strategy   
4        0      10.0        0  Free       0       Everyone         Business   

     Last_Updated         Android_Ver  
0   July 28, 2017          4.1 and up  
1  April 15, 2016          3.0 and up  
2   July 11, 2018          4.2 and up  
3   June 28, 2018  Varies with device  
4  August 6, 2018          4.1 and up

## group by

In [10]:
df.groupby('Category').count()

App  Rating  Reviews  Size_MBs  Installs  Type  Price  \
Category                                                                      
ART_AND_DESIGN         65      62       65        65        65    65     65   
AUTO_AND_VEHICLES      85      73       85        85        85    85     85   
BEAUTY                 53      42       53        53        53    53     53   
BOOKS_AND_REFERENCE   231     178      231       231       231   231    231   
BUSINESS              460     303      460       460       460   460    460   
COMICS                 60      58       60        60        60    60     60   
COMMUNICATION         387     328      387       387       387   387    387   
DATING                234     195      234       234       234   234    234   
EDUCATION             156     155      156       156       156   156    156   
ENTERTAINMENT         149     149      149       149       149   149    149   
EVENTS                 64      45       64        64        64    64     64   
FAMILY               1972    1747     1972      1972      1972  1971   1972   
FINANCE               366     323      366       366       366   366    366   
FOOD_AND_DRINK        127     109      127       127       127   127    127   
GAME                 1144    1097     1144      1144      1144  1144   1144   
HEALTH_AND_FITNESS    341     297      341       341       341   341    341   
HOUSE_AND_HOME         88      76       88        88        88    88     88   
LIBRARIES_AND_DEMO     85      65       85        85        85    85     85   
LIFESTYLE             383     315      383       383       383   383    383   
MAPS_AND_NAVIGATION   137     124      137       137       137   137    137   
MEDICAL               463     350      463       463       463   463    463   
NEWS_AND_MAGAZINES    283     233      283       283       283   283    283   
PARENTING              60      50       60        60        60    60     60   
PERSONALIZATION       392     314      392       392       392   392    392   
PHOTOGRAPHY           335     317      335       335       335   335    335   
PRODUCTIVITY          424     351      424       424       424   424    424   
SHOPPING              260     238      260       260       260   260    260   
SOCIAL                295     259      295       295       295   295    295   
SPORTS                384     319      384       384       384   384    384   
TOOLS                 843     734      843       843       843   843    843   
TRAVEL_AND_LOCAL      258     226      258       258       258   258    258   
VIDEO_PLAYERS         175     160      175       175       175   175    175   
WEATHER                82      75       82        82        82    82     82   

                     Content_Rating  Genres  Last_Updated  Android_Ver  
Category                                                                
ART_AND_DESIGN                   65      65            65           65  
AUTO_AND_VEHICLES                85      85            85           85  
BEAUTY                           53      53            53           53  
BOOKS_AND_REFERENCE             231     231           231          231  
BUSINESS                        460     460           460          460  
COMICS                           60      60            60           60  
COMMUNICATION                   387     387           387          387  
DATING                          234     234           234          234  
EDUCATION                       156     156           156          156  
ENTERTAINMENT                   149     149           149          149  
EVENTS                           64      64            64           64  
FAMILY                         1972    1972          1972         1972  
FINANCE                         366     366           366          366  
FOOD_AND_DRINK                  127     127           127          127  
GAME                           1144    1144          1144         1144  
HEALTH_AND_FITN

In [11]:
df.groupby(['Category','Type']).count()

App  Rating  Reviews  Size_MBs  Installs  Price  \
Category          Type                                                    
ART_AND_DESIGN    Free   62      59       62        62        62     62   
                  Paid    3       3        3         3         3      3   
AUTO_AND_VEHICLES Free   82      72       82        82        82     82   
                  Paid    3       1        3         3         3      3   
BEAUTY            Free   53      42       53        53        53     53   
...                     ...     ...      ...       ...       ...    ...   
TRAVEL_AND_LOCAL  Paid   12       8       12        12        12     12   
VIDEO_PLAYERS     Free  171     156      171       171       171    171   
                  Paid    4       4        4         4         4      4   
WEATHER           Free   74      68       74        74        74     74   
                  Paid    8       7        8         8         8      8   

                        Content_Rating  Genres  Last_Updated  Android_Ver  
Category          Type                                                     
ART_AND_DESIGN    Free              62      62            62           62  
                  Paid               3       3             3            3  
AUTO_AND_VEHICLES Free              82      82            82           82  
                  Paid               3       3             3            3  
BEAUTY            Free              53      53            53           53  
...                                ...     ...           ...          ...  
TRAVEL_AND_LOCAL  Paid              12      12            12           12  
VIDEO_PLAYERS     Free             171     171           171          171  
                  Paid               4       4             4            4  
WEATHER           Free              74      74            74           74  
                  Paid               8       8             8            8  

[63 rows x 10 columns]

In [12]:
# df.groupby(['Category','Type']).count().index

## To find if there are any missing/null values

In [13]:
df.isna().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size_MBs             0
Installs             0
Type                 1
Price                0
Content_Rating       0
Genres               0
Last_Updated         0
Android_Ver          2
dtype: int64

In [14]:
df['Rating'].isnull().values.any()

True

### We can see Ratings has a total of 1474 NaN values, Type has 1 and Android_Ver has 2 NaN values.

- If we have very less number of values we can drop those rows, it won't affect the data drastically

- But if we have a large number of missing values, dropping them wouldn't be an ideal approach. Instead we can fill these null values.

- There are different methods, to fill these values and it completely depends on the type of data we are dealing with.
    - Fill missing values with 0.
    - Fill them with either maximum of the exisitng values or maybe average.

- In this data, we can fill the Ratings with average Ratings grouped by **Category** of the app.

In [27]:
df['Rating'] = df.groupby('Category')['Rating'].transform(lambda x : x.fillna(x.mean()))

In [26]:
df.groupby('Category')['Rating'].transform(lambda x : x.fillna(x.mean()))

0        4.255598
1        4.192272
2        4.335987
3        4.192272
4        4.121452
           ...   
10836    4.500000
10837    4.500000
10838    4.500000
10839    4.500000
10840    4.500000
Name: Rating, Length: 10841, dtype: float64

## To find unique values of a particular feature

In [31]:
df['Content_Rating'].unique()

array(['Teen', 'Everyone', 'Everyone 10+', 'Mature 17+', 'Unrated',
       'Adults only 18+'], dtype=object)

In [32]:
df['Content_Rating'].nunique()

6

In [33]:
df['Content_Rating'].value_counts()

Everyone           8715
Teen               1208
Mature 17+          499
Everyone 10+        414
Adults only 18+       3
Unrated               2
Name: Content_Rating, dtype: int64

### Accessing particular rows

In [34]:
df.head()

App         Category    Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL  4.255598   
1                    Ain Arabic Kids Alif Ba ta           FAMILY  4.192272   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION  4.335987   
3                     Command & Conquer: Rivals           FAMILY  4.192272   
4                                    CX Network         BUSINESS  4.121452   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  \
0        0       8.7        0  Paid  $13.99           Teen           Social   
1        0      33.0        0  Paid   $2.99       Everyone        Education   
2        0       5.5        0  Paid   $1.49       Everyone  Personalization   
3        0      19.0        0   NaN       0   Everyone 10+         Strategy   
4        0      10.0        0  Free       0       Everyone         Business   

     Last_Updated         Android_Ver  
0   July 28, 2017          4.1 and up  
1  April 15, 2016          3.0 and up  
2   July 11, 2018          4.2 and up  
3   June 28, 2018  Varies with device  
4  August 6, 2018          4.1 and up

In [35]:
df.iloc[2]

App               Popsicle Launcher for Android P 9.0 launcher
Category                                       PERSONALIZATION
Rating                                                4.335987
Reviews                                                      0
Size_MBs                                                   5.5
Installs                                                     0
Type                                                      Paid
Price                                                    $1.49
Content_Rating                                        Everyone
Genres                                         Personalization
Last_Updated                                     July 11, 2018
Android_Ver                                         4.2 and up
Name: 2, dtype: object

In [36]:
df.iloc[2:4]

App         Category    Rating  \
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION  4.335987   
3                     Command & Conquer: Rivals           FAMILY  4.192272   

   Reviews  Size_MBs Installs  Type  Price Content_Rating           Genres  \
2        0       5.5        0  Paid  $1.49       Everyone  Personalization   
3        0      19.0        0   NaN      0   Everyone 10+         Strategy   

    Last_Updated         Android_Ver  
2  July 11, 2018          4.2 and up  
3  June 28, 2018  Varies with device

In [43]:
df = df.set_index('App')

In [44]:
df.iloc[4]

Category                BUSINESS
Rating                  4.121452
Reviews                        0
Size_MBs                    10.0
Installs                       0
Type                        Free
Price                          0
Content_Rating          Everyone
Genres                  Business
Last_Updated      August 6, 2018
Android_Ver           4.1 and up
Name: CX Network, dtype: object

In [45]:
df.loc['CX Network']

Category                BUSINESS
Rating                  4.121452
Reviews                        0
Size_MBs                    10.0
Installs                       0
Type                        Free
Price                          0
Content_Rating          Everyone
Genres                  Business
Last_Updated      August 6, 2018
Android_Ver           4.1 and up
Name: CX Network, dtype: object

In [48]:
df = df.reset_index()

## Conditional Filtering

In [51]:
df_tips = pd.read_csv('./tips.csv')

In [52]:
df_tips.tail()

total_bill   tip     sex smoker   day    time  size  price_per_person  \
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
239    Michael Avila  5296068606052842    Sat2657  
240   Monica Sanders  3506806155565404    Sat1766  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

In [55]:
rows_only_male = df_tips[df_tips['sex']=='Male']

rows_only_male.count()

total_bill          157
tip                 157
sex                 157
smoker              157
day                 157
time                157
size                157
price_per_person    157
Payer Name          157
CC Number           157
Payment ID          157
dtype: int64

In [57]:
rows_only_female = df_tips[df_tips['sex'] == 'Female']
rows_only_female.describe().transpose()

count          mean           std           min  \
total_bill         87.0  1.805690e+01  8.009209e+00  3.070000e+00   
tip                87.0  2.833448e+00  1.159495e+00  1.000000e+00   
size               87.0  2.459770e+00  9.376442e-01  1.000000e+00   
price_per_person   87.0  7.464368e+00  2.461255e+00  2.880000e+00   
CC Number          87.0  2.450499e+15  2.326826e+15  6.046330e+10   

                           25%           50%           75%           max  
total_bill        1.275000e+01  1.640000e+01  2.152000e+01  4.430000e+01  
tip               2.000000e+00  2.750000e+00  3.500000e+00  6.500000e+00  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.685000e+00  6.710000e+00  8.725000e+00  1.477000e+01  
CC Number         4.859865e+12  3.508912e+15  4.447818e+15  6.542729e+15

In [59]:
df_tips[(df_tips['price_per_person'] < 4) & (df_tips['sex'] == 'Female')]

total_bill   tip     sex smoker  day    time  size  price_per_person  \
16       10.33  1.67  Female     No  Sun  Dinner     3              3.44   
67        3.07  1.00  Female    Yes  Sat  Dinner     1              3.07   
92        5.75  1.00  Female    Yes  Fri  Dinner     2              2.88   

          Payer Name         CC Number Payment ID  
16  Elizabeth Foster  4240025044626033    Sun9715  
67     Tiffany Brock  4359488526995267    Sat3455  
92      Leah Ramirez  3508911676966392    Fri3780

In [121]:
df_tips[(df_tips['price_per_person'] < 4) | (df_tips['sex'] == 'Female')]

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         33.98  1.01  Female     No   Sun  Dinner     2              8.49   
1         20.68  1.66    Male     No   Sun  Dinner     3              3.45   
4         49.18  3.61  Female     No   Sun  Dinner     4              6.15   
11        70.52  5.00  Female     No   Sun  Dinner     4              8.82   
14        29.66  3.02  Female     No   Sun  Dinner     2              7.42   
..          ...   ...     ...    ...   ...     ...   ...               ...   
226       20.18  2.00  Female    Yes   Fri   Lunch     2              5.04   
229       44.24  2.88  Female    Yes   Sat  Dinner     2             11.06   
238       71.66  4.67  Female     No   Sat  Dinner     3             11.94   
240       54.36  2.00  Female    Yes   Sat  Dinner     2             13.59   
243       37.56  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID last_four_cc  \
0    Christy Cunningham  3560325168603410    Sun2959         3410   
1        Douglas Tucker  4478071379779230    Sun4608         9230   
4          Tonya Carter  4832732618637221    Sun2251         7221   
11         Diane Macias  4577817359320969    Sun6686         0969   
14        Vanessa Jones    30016702287574    Sun3848         7574   
..                  ...               ...        ...          ...   
226          Ruth Weiss  5268689490381635    Fri6359         1635   
229    Jennifer Russell     4793003293608    Sat3943         3608   
238      Kimberly Crane      676184013727    Sat9777         3727   
240      Monica Sanders  3506806155565404    Sat1766         5404   
243     Michelle Hardin  3511451626698139    Thur672         8139   

    Quality_bill Quality  
0          Other   Other  
1          Other   Other  
4          Other   Other  
11         Other   Other  
14         Other   Other  
..           ...     ...  
226        Other   Other  
229        Other   Other  
238        Other   Other  
240        Other   Other  
243        Other   Other  

[92 rows x 14 columns]

In [123]:
df_tips.query('price_per_person < 4 and price_per_person > 2')

total_bill   tip     sex smoker   day    time  size  price_per_person  \
1         20.68  1.66    Male     No   Sun  Dinner     3              3.45   
16        20.66  1.67  Female     No   Sun  Dinner     3              3.44   
67         6.14  1.00  Female    Yes   Sat  Dinner     1              3.07   
92        11.50  1.00  Female    Yes   Fri  Dinner     2              2.88   
149       15.02  2.00    Male     No  Thur   Lunch     2              3.76   
172       14.50  5.15    Male    Yes   Sun  Dinner     2              3.62   
195       15.12  1.44    Male     No  Thur   Lunch     2              3.78   
218       15.48  1.44    Male    Yes   Sat  Dinner     2              3.87   

           Payer Name         CC Number Payment ID last_four_cc Quality_bill  \
1      Douglas Tucker  4478071379779230    Sun4608         9230        Other   
16   Elizabeth Foster  4240025044626033    Sun9715         6033        Other   
67      Tiffany Brock  4359488526995267    Sat3455         5267        Other   
92       Leah Ramirez  3508911676966392    Fri3780         6392        Other   
149    Daniel Robbins  4823139288341889   Thur6321         1889        Other   
172       Larry White    30432617123103    Sun9209         3103     Generous   
195     Michael White  4865390263095532    Thur697         5532        Other   
218   Nicholas Archer   340517153733524    Sat4772         3524        Other   

      Quality  
1       Other  
16      Other  
67      Other  
92      Other  
149     Other  
172  Generous  
195     Other  
218     Other

In [127]:
gender = 'Female'
df_tips.query('price_per_person < 4 and sex == @gender')

total_bill   tip     sex smoker  day    time  size  price_per_person  \
16       20.66  1.67  Female     No  Sun  Dinner     3              3.44   
67        6.14  1.00  Female    Yes  Sat  Dinner     1              3.07   
92       11.50  1.00  Female    Yes  Fri  Dinner     2              2.88   

          Payer Name         CC Number Payment ID last_four_cc Quality_bill  \
16  Elizabeth Foster  4240025044626033    Sun9715         6033        Other   
67     Tiffany Brock  4359488526995267    Sat3455         5267        Other   
92      Leah Ramirez  3508911676966392    Fri3780         6392        Other   

   Quality  
16   Other  
67   Other  
92   Other

In [62]:
# df_tips[(df_tips['day'] == "Sun") | (df_tips['day'] == 'Sat')]
df_tips[df_tips['day'].isin(['Sat','Sun'])]

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0         16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No  Sun  Dinner     4              6.15   
..          ...   ...     ...    ...  ...     ...   ...               ...   
238       35.83  4.67  Female     No  Sat  Dinner     3             11.94   
239       29.03  5.92    Male     No  Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes  Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes  Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No  Sat  Dinner     2              8.91   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
238      Kimberly Crane      676184013727    Sat9777  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  

[163 rows x 11 columns]

### Useful methods on single column

In [63]:
def last_four(num):
    return str(num)[-4:]

In [65]:
df_tips['last_four_cc'] = df_tips['CC Number'].apply(last_four)

In [66]:
df_tips['total_bill'] = df_tips['total_bill'].apply(lambda x : x*2)

In [68]:
df_tips.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       33.98  1.01  Female     No  Sun  Dinner     2              8.49   
1       20.68  1.66    Male     No  Sun  Dinner     3              3.45   
2       42.02  3.50    Male     No  Sun  Dinner     3              7.00   
3       47.36  3.31    Male     No  Sun  Dinner     2             11.84   
4       49.18  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID last_four_cc  
0  Christy Cunningham  3560325168603410    Sun2959         3410  
1      Douglas Tucker  4478071379779230    Sun4608         9230  
2      Travis Walters  6011812112971322    Sun4458         1322  
3    Nathaniel Harris  4676137647685994    Sun5260         5994  
4        Tonya Carter  4832732618637221    Sun2251         7221

### Useful methods on multiple columns

In [69]:
def quality_bill(total_bill,tip):
    if(tip/total_bill > 0.25):
        return 'Generous'
    return 'Other'

In [70]:
quality_bill(8.77,2)

'Other'

In [71]:
df_tips['Quality_bill'] = df_tips[['total_bill','tip']].apply(lambda df_tips : quality_bill(df_tips['total_bill'],df_tips['tip']),axis=1)

In [73]:
# np.vectorize() awares the function about the DataFrame.
df_tips['Quality']  = np.vectorize(quality_bill)(df_tips['total_bill'],df_tips['tip'])

In [75]:
df_tips.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       33.98  1.01  Female     No  Sun  Dinner     2              8.49   
1       20.68  1.66    Male     No  Sun  Dinner     3              3.45   
2       42.02  3.50    Male     No  Sun  Dinner     3              7.00   
3       47.36  3.31    Male     No  Sun  Dinner     2             11.84   
4       49.18  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID last_four_cc Quality_bill  \
0  Christy Cunningham  3560325168603410    Sun2959         3410        Other   
1      Douglas Tucker  4478071379779230    Sun4608         9230        Other   
2      Travis Walters  6011812112971322    Sun4458         1322        Other   
3    Nathaniel Harris  4676137647685994    Sun5260         5994        Other   
4        Tonya Carter  4832732618637221    Sun2251         7221        Other   

  Quality  
0   Other  
1   Other  
2   Other  
3   Other  
4   Other

## USEFUL METHODS : STATISTICAL INFORMATION AND SORTING

In [76]:
df.describe().transpose()

count           mean           std       min        25%  \
Rating    10841.0       4.190199  4.805656e-01  1.000000   4.047411   
Reviews   10841.0  444111.928051  2.927629e+06  0.000000  38.000000   
Size_MBs  10841.0      19.774147  2.140435e+01  0.008301   4.900000   

                  50%      75%         max  
Rating       4.259664      4.5         5.0  
Reviews   2094.000000  54768.0  78158306.0  
Size_MBs    11.000000     27.0       100.0

In [80]:
df_tips.sort_values(['tip'])


total_bill    tip     sex smoker   day    time  size  price_per_person  \
67         6.14   1.00  Female    Yes   Sat  Dinner     1              3.07   
236       25.20   1.00    Male    Yes   Sat  Dinner     2              6.30   
92        11.50   1.00  Female    Yes   Fri  Dinner     2              2.88   
111       14.50   1.00  Female     No   Sat  Dinner     1              7.25   
0         33.98   1.01  Female     No   Sun  Dinner     2              8.49   
..          ...    ...     ...    ...   ...     ...   ...               ...   
141       68.60   6.70    Male     No  Thur   Lunch     6              5.72   
59        96.54   6.73    Male     No   Sat  Dinner     4             12.07   
23        78.84   7.58    Male     No   Sat  Dinner     4              9.86   
212       96.66   9.00    Male     No   Sat  Dinner     4             12.08   
170      101.62  10.00    Male    Yes   Sat  Dinner     3             16.94   

             Payer Name         CC Number Payment ID last_four_cc  \
67        Tiffany Brock  4359488526995267    Sat3455         5267   
236       Matthew Myers  3543676378973965    Sat5032         3965   
92         Leah Ramirez  3508911676966392    Fri3780         6392   
111         Terri Jones  3559221007826887    Sat4801         6887   
0    Christy Cunningham  3560325168603410    Sun2959         3410   
..                  ...               ...        ...          ...   
141      Steven Carlson  3526515703718508   Thur1025         8508   
59          Brian Ortiz  6596453823950595    Sat8139         0595   
23       Lance Peterson  3542584061609808     Sat239         9808   
212     Alex Williamson      676218815212    Sat4590         5212   
170       Gregory Clark  5473850968388236    Sat1954         8236   

    Quality_bill Quality  
67         Other   Other  
236        Other   Other  
92         Other   Other  
111        Other   Other  
0          Other   Other  
..           ...     ...  
141        Other   Other  
59         Other   Other  
23         Other   Other  
212        Other   Other  
170        Other   Other  

[244 rows x 14 columns]

In [81]:
df_tips.sort_values(['tip','size'])

total_bill    tip     sex smoker   day    time  size  price_per_person  \
67         6.14   1.00  Female    Yes   Sat  Dinner     1              3.07   
111       14.50   1.00  Female     No   Sat  Dinner     1              7.25   
92        11.50   1.00  Female    Yes   Fri  Dinner     2              2.88   
236       25.20   1.00    Male    Yes   Sat  Dinner     2              6.30   
0         33.98   1.01  Female     No   Sun  Dinner     2              8.49   
..          ...    ...     ...    ...   ...     ...   ...               ...   
141       68.60   6.70    Male     No  Thur   Lunch     6              5.72   
59        96.54   6.73    Male     No   Sat  Dinner     4             12.07   
23        78.84   7.58    Male     No   Sat  Dinner     4              9.86   
212       96.66   9.00    Male     No   Sat  Dinner     4             12.08   
170      101.62  10.00    Male    Yes   Sat  Dinner     3             16.94   

             Payer Name         CC Number Payment ID last_four_cc  \
67        Tiffany Brock  4359488526995267    Sat3455         5267   
111         Terri Jones  3559221007826887    Sat4801         6887   
92         Leah Ramirez  3508911676966392    Fri3780         6392   
236       Matthew Myers  3543676378973965    Sat5032         3965   
0    Christy Cunningham  3560325168603410    Sun2959         3410   
..                  ...               ...        ...          ...   
141      Steven Carlson  3526515703718508   Thur1025         8508   
59          Brian Ortiz  6596453823950595    Sat8139         0595   
23       Lance Peterson  3542584061609808     Sat239         9808   
212     Alex Williamson      676218815212    Sat4590         5212   
170       Gregory Clark  5473850968388236    Sat1954         8236   

    Quality_bill Quality  
67         Other   Other  
111        Other   Other  
92         Other   Other  
236        Other   Other  
0          Other   Other  
..           ...     ...  
141        Other   Other  
59         Other   Other  
23         Other   Other  
212        Other   Other  
170        Other   Other  

[244 rows x 14 columns]

In [82]:
df_tips.sort_values('tip', ascending=False)

total_bill    tip     sex smoker   day    time  size  price_per_person  \
170      101.62  10.00    Male    Yes   Sat  Dinner     3             16.94   
212       96.66   9.00    Male     No   Sat  Dinner     4             12.08   
23        78.84   7.58    Male     No   Sat  Dinner     4              9.86   
59        96.54   6.73    Male     No   Sat  Dinner     4             12.07   
141       68.60   6.70    Male     No  Thur   Lunch     6              5.72   
..          ...    ...     ...    ...   ...     ...   ...               ...   
0         33.98   1.01  Female     No   Sun  Dinner     2              8.49   
236       25.20   1.00    Male    Yes   Sat  Dinner     2              6.30   
111       14.50   1.00  Female     No   Sat  Dinner     1              7.25   
67         6.14   1.00  Female    Yes   Sat  Dinner     1              3.07   
92        11.50   1.00  Female    Yes   Fri  Dinner     2              2.88   

             Payer Name         CC Number Payment ID last_four_cc  \
170       Gregory Clark  5473850968388236    Sat1954         8236   
212     Alex Williamson      676218815212    Sat4590         5212   
23       Lance Peterson  3542584061609808     Sat239         9808   
59          Brian Ortiz  6596453823950595    Sat8139         0595   
141      Steven Carlson  3526515703718508   Thur1025         8508   
..                  ...               ...        ...          ...   
0    Christy Cunningham  3560325168603410    Sun2959         3410   
236       Matthew Myers  3543676378973965    Sat5032         3965   
111         Terri Jones  3559221007826887    Sat4801         6887   
67        Tiffany Brock  4359488526995267    Sat3455         5267   
92         Leah Ramirez  3508911676966392    Fri3780         6392   

    Quality_bill Quality  
170        Other   Other  
212        Other   Other  
23         Other   Other  
59         Other   Other  
141        Other   Other  
..           ...     ...  
0          Other   Other  
236        Other   Other  
111        Other   Other  
67         Other   Other  
92         Other   Other  

[244 rows x 14 columns]

In [83]:
# Finding the index location of Max or Min value
# df_tips['total_bill'].max() -> 50.81
df_tips['total_bill'].idxmax()

170

In [87]:
df_tips.iloc[170]

total_bill                    101.62
tip                             10.0
sex                             Male
smoker                           Yes
day                              Sat
time                          Dinner
size                               3
price_per_person               16.94
Payer Name             Gregory Clark
CC Number           5473850968388236
Payment ID                   Sat1954
last_four_cc                    8236
Quality_bill                   Other
Quality                        Other
Name: 170, dtype: object

In [96]:
# Find Correlation -> only for numerical data
df_tips.corr()

total_bill       tip      size  price_per_person  CC Number
total_bill          1.000000  0.675734  0.598315          0.647554   0.104576
tip                 0.675734  1.000000  0.489299          0.347405   0.110857
size                0.598315  0.489299  1.000000         -0.175359  -0.030239
price_per_person    0.647554  0.347405 -0.175359          1.000000   0.135240
CC Number           0.104576  0.110857 -0.030239          0.135240   1.000000

In [97]:
# top ten rows ranked by highest tip
df_tips.nlargest(10,'tip')
# Similar to, -> df_tips.sort_values('tip').iloc[0:10]

total_bill    tip     sex smoker   day    time  size  price_per_person  \
170      101.62  10.00    Male    Yes   Sat  Dinner     3             16.94   
212       96.66   9.00    Male     No   Sat  Dinner     4             12.08   
23        78.84   7.58    Male     No   Sat  Dinner     4              9.86   
59        96.54   6.73    Male     No   Sat  Dinner     4             12.07   
141       68.60   6.70    Male     No  Thur   Lunch     6              5.72   
183       46.34   6.50    Male    Yes   Sun  Dinner     4              5.79   
214       56.34   6.50  Female    Yes   Sat  Dinner     3              9.39   
47        64.80   6.00    Male     No   Sun  Dinner     4              8.10   
239       58.06   5.92    Male     No   Sat  Dinner     3              9.68   
88        49.42   5.85    Male     No  Thur   Lunch     2             12.36   

            Payer Name         CC Number Payment ID last_four_cc Quality_bill  \
170      Gregory Clark  5473850968388236    Sat1954         8236        Other   
212    Alex Williamson      676218815212    Sat4590         5212        Other   
23      Lance Peterson  3542584061609808     Sat239         9808        Other   
59         Brian Ortiz  6596453823950595    Sat8139         0595        Other   
141     Steven Carlson  3526515703718508   Thur1025         8508        Other   
183  Dr. Michael James     4718501859162    Sun6059         9162        Other   
214    Marissa Jackson  4922302538691962    Sat3374         1962        Other   
47        James Barnes  3552002592874186    Sun9677         4186        Other   
239      Michael Avila  5296068606052842    Sat2657         2842        Other   
88        Roger Taylor     4410248629955   Thur9003         9955        Other   

    Quality  
170   Other  
212   Other  
23    Other  
59    Other  
141   Other  
183   Other  
214   Other  
47    Other  
239   Other  
88    Other

## Between

In [92]:
df_tips['total_bill'].between(10,25,'both')

0      False
1       True
2      False
3      False
4      False
       ...  
239    False
240    False
241    False
242    False
243    False
Name: total_bill, Length: 244, dtype: bool

In [94]:
# Only those rows, where 10<=total_bill<=25
df_tips[df_tips['total_bill'].between(10,25,'both')]

total_bill   tip     sex smoker   day    time  size  price_per_person  \
1         20.68  1.66    Male     No   Sun  Dinner     3              3.45   
6         17.54  2.00    Male     No   Sun  Dinner     2              4.38   
10        20.54  1.71    Male     No   Sun  Dinner     2              5.14   
16        20.66  1.67  Female     No   Sun  Dinner     3              3.44   
30        19.10  1.45    Male     No   Sat  Dinner     2              4.78   
43        19.36  1.32    Male     No   Sun  Dinner     2              4.84   
51        20.58  2.60  Female     No   Sun  Dinner     2              5.14   
53        19.88  1.56    Male     No   Sun  Dinner     2              4.97   
58        22.48  1.76    Male    Yes   Sat  Dinner     2              5.62   
62        22.04  1.98    Male    Yes   Sat  Dinner     2              5.51   
70        24.04  1.97    Male     No   Sat  Dinner     2              6.01   
75        21.02  1.25    Male     No   Sat  Dinner     2              5.26   
82        20.14  1.83  Female     No  Thur   Lunch     1             10.07   
92        11.50  1.00  Female    Yes   Fri  Dinner     2              2.88   
97        24.06  1.50    Male    Yes   Fri  Dinner     2              6.02   
99        24.92  1.50    Male     No   Fri  Dinner     2              6.23   
100       22.70  2.50  Female    Yes   Fri  Dinner     2              5.68   
111       14.50  1.00  Female     No   Sat  Dinner     1              7.25   
117       21.30  1.50  Female     No  Thur   Lunch     2              5.32   
118       24.86  1.80  Female     No  Thur   Lunch     2              6.22   
120       23.38  2.31    Male     No  Thur   Lunch     2              5.84   
124       24.96  2.52  Female     No  Thur   Lunch     2              6.24   
126       17.04  1.48    Male     No  Thur   Lunch     2              4.26   
128       22.76  2.00  Female     No  Thur   Lunch     2              5.69   
132       22.34  1.50  Female     No  Thur   Lunch     2              5.58   
133       24.52  2.00  Female     No  Thur   Lunch     2              6.13   
135       17.02  1.25  Female     No  Thur   Lunch     2              4.26   
136       20.66  2.00  Female     No  Thur   Lunch     2              5.16   
145       16.70  1.50  Female     No  Thur   Lunch     2              4.18   
147       23.74  1.63  Female     No  Thur   Lunch     2              5.94   
148       19.56  1.73    Male     No  Thur   Lunch     2              4.89   
149       15.02  2.00    Male     No  Thur   Lunch     2              3.76   
168       21.18  1.61  Female    Yes   Sat  Dinner     2              5.30   
169       21.26  2.00  Female    Yes   Sat  Dinner     2              5.32   
172       14.50  5.15    Male    Yes   Sun  Dinner     2              3.62   
178       19.20  4.00  Female    Yes   Sun  Dinner     2              4.80   
195       15.12  1.44    Male     No  Thur   Lunch     2              3.78   
196       20.68  2.00    Male    Yes  Thur   Lunch     2              5.17   
217       23.18  1.50    Male    Yes   Sat  Dinner     2              5.80   
218       15.48  1.44    Male    Yes   Sat  Dinner     2              3.87   
220       24.32  2.20    Male    Yes   Fri   Lunch     2              6.08   
222       17.16  1.92    Male    Yes   Fri   Lunch     1              8.58   
226       20.18  2.00  Female    Yes   Fri   Lunch     2              5.04   
232       23.22  3.39    Male     No   Sat  Dinner     2              5.80   
233       21.54  1.47    Male     No   Sat  Dinner     2              5.38   
235       20.14  1.25    Male     No   Sat  Dinner     2              5.04   

              Payer Name         CC Number Payment ID last_four_cc  \
1         Douglas Tucker  4478071379779230    Sun4608         9230   
6     Kristopher Johnson  2223727524230344    Sun5985         0344   
10         William Riley      566287581219    Sun2546         1219   
16      Elizabeth Foster  4240025044626033    Sun9715     

## Replace

In [88]:
df_tips['Quality_bill'].replace('Other',"Others")

0      Others
1      Others
2      Others
3      Others
4      Others
        ...  
239    Others
240    Others
241    Others
242    Others
243    Others
Name: Quality_bill, Length: 244, dtype: object

In [89]:
# df_tips['sex'].replace("Male",'M')
df_tips['sex'].replace(['Male','Female'],['M','F'])

0      F
1      M
2      M
3      M
4      F
      ..
239    M
240    F
241    M
242    M
243    F
Name: sex, Length: 244, dtype: object

In [90]:
# Replace by using map()
mappingDict = {'Female':'F','Male':'M'}
df_tips['sex'].map(mappingDict)

0      F
1      M
2      M
3      M
4      F
      ..
239    M
240    F
241    M
242    M
243    F
Name: sex, Length: 244, dtype: object

## **CONCATENATION**

In [98]:
data_one = {'A':['A0','A1','A2','A3'],'B':['B0','B1','B2','B3']}
data_two = {'C':['C0','C1','C2','C3'],'D':['D0','D1','D2','D3']}

In [99]:
one = pd.DataFrame(data_one)
two = pd.DataFrame(data_two)

In [102]:
print(one)
print(two)

    A   B
0  A0  B0
1  A1  B1
2  A2  B2
3  A3  B3
    C   D
0  C0  D0
1  C1  D1
2  C2  D2
3  C3  D3


In [103]:
pd.concat([one, two])

A    B    C    D
0   A0   B0  NaN  NaN
1   A1   B1  NaN  NaN
2   A2   B2  NaN  NaN
3   A3   B3  NaN  NaN
0  NaN  NaN   C0   D0
1  NaN  NaN   C1   D1
2  NaN  NaN   C2   D2
3  NaN  NaN   C3   D3

In [105]:
pd.concat([one, two], axis=1)

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

In [106]:
two.columns = one.columns

two

A   B
0  C0  D0
1  C1  D1
2  C2  D2
3  C3  D3

In [107]:
pd.concat([one,two],axis=0)


A   B
0  A0  B0
1  A1  B1
2  A2  B2
3  A3  B3
0  C0  D0
1  C1  D1
2  C2  D2
3  C3  D3

## **MERGE**

In [108]:
# MERGE is analogous to JOIN in SQL
# Three ways to merge tables : Inner, Outer and (left/right)

In [109]:
registrations = pd.DataFrame({"reg_id":[1,2,3,4],"Name":['Andrew','Blob','Casie',"Daniel"]})
logins = pd.DataFrame({"login_id":[1,2,3,4],"Name":['Cassandra','Daniel','Xavier',"Blob"]})

In [110]:
print(registrations)
print(logins)

   reg_id    Name
0       1  Andrew
1       2    Blob
2       3   Casie
3       4  Daniel
   login_id       Name
0         1  Cassandra
1         2     Daniel
2         3     Xavier
3         4       Blob


In [111]:
pd.merge(registrations,logins,how='inner',on='Name')

reg_id    Name  login_id
0       2    Blob         4
1       4  Daniel         2

In [112]:
pd.merge(registrations,logins,how='left',on='Name')

reg_id    Name  login_id
0       1  Andrew       NaN
1       2    Blob       4.0
2       3   Casie       NaN
3       4  Daniel       2.0

In [113]:
pd.merge(registrations,logins,how='right',on='Name')

reg_id       Name  login_id
0     NaN  Cassandra         1
1     4.0     Daniel         2
2     NaN     Xavier         3
3     2.0       Blob         4

In [114]:
pd.merge(registrations,logins,how='outer',on='Name')

reg_id       Name  login_id
0     1.0     Andrew       NaN
1     2.0       Blob       4.0
2     3.0      Casie       NaN
3     4.0     Daniel       2.0
4     NaN  Cassandra       1.0
5     NaN     Xavier       3.0

In [115]:
registrations =  registrations.set_index('Name')

In [116]:
# If we want to merge with Index of one table with column of another table
pd.merge(left=registrations,right=logins,how='inner',left_index=True,right_on='Name')

reg_id  login_id    Name
3       2         4    Blob
1       4         2  Daniel

In [117]:
registrations = registrations.reset_index()

In [118]:
 # If column names are different on which we have to apply the merge
registrations.columns = ['reg_name',"reg_id"]

In [119]:
registrations

reg_name  reg_id
0   Andrew       1
1     Blob       2
2    Casie       3
3   Daniel       4

In [120]:
pd.merge(registrations,logins,how='inner',left_on='reg_name',right_on='Name')

reg_name  reg_id  login_id    Name
0     Blob       2         4    Blob
1   Daniel       4         2  Daniel